apis/infrastructure/v1beta1/bootstrapkubeconfig_webhook.go

```diff
@@ -0,0 +1,102 @@
+// Copyright 2022 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package v1beta1
+
+import (
+	b64 "encoding/base64"
+	"encoding/pem"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	ctrl "sigs.k8s.io/controller-runtime"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+// log is for logging in this package.
+var bootstrapkubeconfiglog = logf.Log.WithName("bootstrapkubeconfig-resource")
+
+func (r *BootstrapKubeconfig) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(r).
+		Complete()
+}
+
+//+kubebuilder:webhook:path=/validate-infrastructure-cluster-x-k8s-io-v1beta1-bootstrapkubeconfig,mutating=false,failurePolicy=fail,sideEffects=None,groups=infrastructure.cluster.x-k8s.io,resources=bootstrapkubeconfigs,verbs=create;update,versions=v1beta1,name=vbootstrapkubeconfig.kb.io,admissionReviewVersions=v1
+
+var _ webhook.Validator = &BootstrapKubeconfig{}
+
+// APIServerSplitStringLength indicates that after we split the APIServer string on ':',
+// there will be 3 parts to it
+const APIServerSplitStringLength = 3
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
+func (r *BootstrapKubeconfig) ValidateCreate() error {
+	bootstrapkubeconfiglog.Info("validate create", "name", r.Name)
+
+	if err := r.validateAPIServer(); err != nil {
+		return err
+	}
+
+	if err := r.validateCAData(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type
+func (r *BootstrapKubeconfig) ValidateUpdate(old runtime.Object) error {
+	bootstrapkubeconfiglog.Info("validate update", "name", r.Name)
+
+	if err := r.validateAPIServer(); err != nil {
+		return err
+	}
+
+	if err := r.validateCAData(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// ValidateDelete implements webhook.Validator so a webhook will be registered for the type
+func (r *BootstrapKubeconfig) ValidateDelete() error {
+	bootstrapkubeconfiglog.Info("validate delete", "name", r.Name)
+
+	// TODO(user): fill in your validation logic upon object update.
+	return nil
+}
+
+func (r *BootstrapKubeconfig) validateAPIServer() error {
+	if r.Spec.APIServer == "" {
+		return field.Invalid(field.NewPath("spec").Child("apiserver"), r.Spec.APIServer, "APIServer field cannot be empty")
+	}
+
+	// this is a basic check to see if the APIServer is of the format https://hostname:port
+	// will add more regex if needed
+	if len(strings.Split(r.Spec.APIServer, ":")) != APIServerSplitStringLength {
```

Comment:
Could we use something from the standard library? this maybe https://pkg.go.dev/net/url#Parse

Reply:
yup. thanks. much better.